app/app.js

```diff
@@ -364,43 +404,48 @@ app.post('/extract', [
                   return el.getAttribute(fnAttribute);
                 }, pdfElement, fnAttribute);
               }
-              if (!pdfLink) {
-                throw new Error(`Element ${fnElement} with attribute ${fnAttribute} not found in ${fnUrl}`);
-              }
-
-              log.info(lgParams, `Extracted ${fnElement} from ${fnUrl} with attribute ${fnAttribute} and value ${pdfLink}`);
 
               // Grab the file as a blob if requested.
               if (fnFile) {
+                if (fnElement2) {
+                  pdfElement = await page.$(fnElement2);
+                  if (pdfElement) {
+                    pdfLink = await page.evaluate((el, fnAttribute) => {
+                      return el.getAttribute(fnAttribute);
+                    }, pdfElement, fnAttribute);
+                  }
+                }
+
                 let fileName = path.basename(pdfLink);
                 if (!fileName) {
                   fileName = `downloaded-${Date.now()}.pdf`;
                 }
-                const filePath = path.resolve(downloadPath, fileName);
-
-                if (!pdfLink.startsWith('http')) {
-                  // Use hostname from the URL.
-                  const urlObj = new URL(fnUrl);
-                  const hostname = urlObj.hostname;
-                  const port = urlObj.port ? `:${urlObj.port}` : '';
-                  const protocol = urlObj.protocol;
-                  const baseUrl = `${protocol}//${hostname}${port}`;
-                  pdfLink = `${baseUrl}${pdfLink}`;
-                }
+                let filePath = path.resolve(downloadPath, fileName);
 
                 try {
-                    // Download the file.
-                    await fetch(pdfLink).then(response => response.blob())
-                      .then(blob => blob.arrayBuffer())
-                      .then(Buffer.from)
-                      .then((buf) => {
-                          fs.writeFileSync(filePath, buf);
-                          console.log('file written');
-                      });
-
-                    console.log(`Downloaded: ${fileName} from ${pdfLink} and saved to ${filePath}`);
-
-                    // Read the file
+                    // Use puppeteer to download file.
+                    await sleep(444);
+                    await page.click(fnElement);
+                    await sleep(555);
+
+                    // Use second element if provided.
+                    if (fnElement2) {
+                      await page.waitForSelector(fnElement2);
+                      await page.click(fnElement2);
+                      await sleep(666);
+                    }
+
+                    // Check every 500ms.
+                    await new Promise((resolve) => {
```

Comment:
Should there be some timeout to avoid possible long loops?

Reply:
Yes, added

PS: default is 2 minutes